{"cells":[{"cell_type":"markdown","metadata":{},"source":"# Preprocessing and clustering functions"},{"cell_type":"markdown","metadata":{},"source":"This tutorial explores preprocessing and clustering"},{"cell_type":"markdown","metadata":{},"source":"## Scatter plots for embeddings"},{"cell_type":"markdown","metadata":{},"source":"With scanpy, scatter plots for tSNE, UMAP and several other embeddings are readily available using the `sc.pl.tsne`, `sc.pl.umap` etc. functions. See [here](https://scanpy.readthedocs.io/en/stable/api/scanpy.plotting.html#embeddings) the list of options.\n\nThose functions access the data stored in `adata.obsm`. For example `sc.pl.umap` uses the information stored in `adata.obsm['X_umap']`. For more flexibility, any key stored in `adata.obsm` can be used with the generic function `sc.pl.embedding`."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"import scanpy as sc\nimport pandas as pd\nfrom matplotlib.pyplot import rc_context\nimport scipy\nimport anndata\nimport os\nimport warnings\nwarnings.filterwarnings(\"ignore\")"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"%matplotlib inline\nsc.set_figure_params(dpi=100, color_map = 'viridis_r')\nsc.settings.verbosity = 3\nsc.logging.print_header()"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# configuration for capsule\nfrom pathlib import Path\n\nsc._settings.ScanpyConfig.cachedir = Path(\"/data/cache/\")\nsc._settings.ScanpyConfig.datasetdir = Path(\"/data/\")\nPath(\"/results/figures\").mkdir(parents=True, exist_ok=True)\nsc._settings.ScanpyConfig.figdir = Path(\"/results/figures\")\nPath(\"/results/write\").mkdir(parents=True, exist_ok=True)\nsc._settings.ScanpyConfig.writedir = Path(\"/results/write/\")"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"results_dir = \"/results\"\nresults_file = (\n    f\"{results_dir}/write/INCF.h5ad\"  # the file that will store the analysis results\n)"},{"cell_type":"markdown","metadata":{},"source":"### Load dataset"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"#adata = scipy.io.mmread(\"/data/input_data/matrix.mtx\")\n#cells = pd.read_csv(\"/data/input_data/barcodes.tsv\",sep=\"\\t\",header=None)\n#genes = pd.read_csv(\"/data/input_data/features.tsv\",sep=\"\\t\",header=None)\n\nadata = sc.read_10x_mtx(\n    '/data/INCF_10X_data/',  # the directory with the `.mtx` file\n    var_names='gene_symbols',                # use gene symbols for the variable names (variables-axis index)\n    cache=True)                              # write a cache file for faster subsequent reading"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata.var_names_make_unique()"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Show those genes that yield the highest fraction of counts in each single cell, across all cells.\n\nsc.pl.highest_expr_genes(\n    adata,\n    n_top=20,\n)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Basic Filtering\n\nsc.pp.filter_cells(adata, min_genes=200)\nsc.pp.filter_genes(adata, min_cells=3)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata.var[\"mt\"] = adata.var_names.str.startswith(\n    \"MT-\"\n)  # annotate the group of mitochondrial genes as 'mt'\nsc.pp.calculate_qc_metrics(\n    adata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True\n)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.violin(\n    adata,\n    [\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n    jitter=0.4,\n    multi_panel=True,\n)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Remove cells that have too many mitochondrial genes expressed or too many total counts:\n\nsc.pl.scatter(adata, x=\"total_counts\", y=\"pct_counts_mt\")\nsc.pl.scatter(adata, x=\"total_counts\", y=\"n_genes_by_counts\")"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Actually do the filtering\n\nadata = adata[adata.obs.n_genes_by_counts < 2500, :]\nadata = adata[adata.obs.pct_counts_mt < 5, :]"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Total-count normalize (library-size correct) the data matrix $\\mathbf{X}$ to 10,000 reads per cell, so that counts become comparable among cells.\n\nsc.pp.normalize_total(adata, target_sum=1e4)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Logarithmize the data:\n\nsc.pp.log1p(adata)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Identify highly-variable genes.\n\nsc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\nsc.pl.highly_variable_genes(adata)"},{"cell_type":"markdown","metadata":{},"source":"Set the `.raw` attribute of the AnnData object to the normalized and logarithmized raw gene expression for later use in differential testing and visualizations of gene expression. This simply freezes the state of the AnnData object."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata.raw = adata"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Actually do the filtering\n\nadata = adata[:, adata.var.highly_variable]"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Regress out effects of total counts per cell and the percentage of mitochondrial genes expressed. Scale the data to unit variance.\n\nsc.pp.regress_out(adata, [\"total_counts\", \"pct_counts_mt\"])"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Scale each gene to unit variance. Clip values exceeding standard deviation 10.\n\nsc.pp.scale(adata, max_value=10)"},{"cell_type":"markdown","metadata":{},"source":"## Principal component analysis"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and denoises the data.\n\nsc.tl.pca(adata, svd_solver=\"arpack\")"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.pca(adata, color='AURKAIP1')"},{"cell_type":"markdown","metadata":{},"source":"Let us inspect the contribution of single PCs to the total variance in the data. This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells, e.g. used in the clustering function  `sc.tl.louvain()` or tSNE `sc.tl.tsne()`. In our experience, often a rough estimate of the number of PCs does fine."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.pca_variance_ratio(adata, log=True)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata.write(results_file)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata"},{"cell_type":"markdown","metadata":{},"source":"## Computing the neighborhood graph"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Let us compute the neighborhood graph of cells using the PCA representation of the data matrix. \n# You might simply use default values here. For the sake of reproducing Seurat's results, let's take the following values.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)"},{"cell_type":"markdown","metadata":{},"source":"## Embedding the neighborhood graph\n\nWe suggest embedding the graph in two dimensions using UMAP ([McInnes et al., 2018](https://arxiv.org/abs/1802.03426)), see below. It is potentially more faithful to the global connectivity of the manifold than tSNE, i.e., it better preserves trajectories. In some ocassions, you might still observe disconnected clusters and similar connectivity violations. They can usually be remedied by running:\n\n```Python\nsc.tl.paga(adata)\nsc.pl.paga(adata, plot=False)  # remove `plot=False` if you want to see the coarse-grained graph\nsc.tl.umap(adata, init_pos=\"paga\")\n```"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.tl.umap(adata)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# adata.var[\"gene_ids\"][10:25]\n"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.umap(adata, color=[\"C1QL3\", \"FOXP2\", \"RORA\"])"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.umap(adata, color=[\"C1QL3\", \"FOXP2\", \"RORA\"], use_raw=False)"},{"cell_type":"markdown","metadata":{},"source":"## Clustering the neighborhood graph"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.tl.leiden(adata)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.umap(adata, color=[\"leiden\", \"C1QL3\", \"FOXP2\", \"RORA\"])"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Save the result\n\nadata.write(results_file)"},{"cell_type":"markdown","metadata":{},"source":"#### Let us compute a ranking for the highly differential genes in each cluster. For this, by default, the `.raw` attribute of AnnData is used in case it has been initialized before. The simplest and fastest method to do so is the t-test."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.tl.rank_genes_groups(adata, 'leiden', method='t-test')\nsc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"},{"cell_type":"markdown","metadata":{},"source":"The result of a [Wilcoxon rank-sum (Mann-Whitney-U)](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test) test is very similar. We recommend using the latter in publications, see e.g., [Sonison & Robinson (2018)](https://doi.org/10.1038/nmeth.4612). You might also consider much more powerful differential testing packages like MAST, limma, DESeq2 and, for python, the recent diffxpy."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon')\nsc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata.write(results_file)"},{"cell_type":"markdown","metadata":{},"source":"As an alternative, let us rank genes using logistic regression. For instance, this has been suggested by [Natranos et al. (2018)](https://doi.org/10.1101/258566). The essential difference is that here, we use a multi-variate appraoch whereas conventional differential tests are uni-variate. [Clark et al. (2014)](https://doi.org/10.1186/1471-2105-15-79) has more details."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.tl.rank_genes_groups(adata, 'leiden', method='logreg')\nsc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Reload the object that has been save with the Wilcoxon Rank-Sum test result.\n\nadata = sc.read(results_file)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Show the 15 top ranked genes per cluster 0, 1, ..., 14 in a dataframe.\n\npd.DataFrame(adata.uns['rank_genes_groups']['names']).head(15)\n"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Get a table with the scores and groups.\n\nresult = adata.uns[\"rank_genes_groups\"]\ngroups = result[\"names\"].dtype.names\npd.DataFrame(\n    {\n        group + \"_\" + key[:1]: result[key][group]\n        for group in groups\n        for key in [\"names\", \"pvals\"]\n    }\n).head(15)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Compare to a single cluster:\n\nsc.tl.rank_genes_groups(adata, \"leiden\", groups=[\"0\"], reference=\"1\", method=\"wilcoxon\")\nsc.pl.rank_genes_groups(adata, groups=[\"0\"], n_genes=20)"},{"cell_type":"markdown","metadata":{},"source":"If we want a more detailed view for a certain group, use `sc.pl.rank_genes_groups_violin`."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.rank_genes_groups_violin(adata, groups=\"0\", n_genes=8)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# Reload the object with the computed differential expression (i.e. DE via a comparison with the rest of the groups):\n\nadata = sc.read(results_file)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"sc.pl.rank_genes_groups_violin(adata, groups=\"0\", n_genes=8)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# If you want to compare a certain gene across groups, use the following.\n\nsc.pl.violin(adata, [\"C1QL3\", \"FOXP2\", \"RORA\"], groupby=\"leiden\")"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata.write(\n    results_file, compression=\"gzip\"\n)  # `compression=\"gzip\"` saves disk space, but slows down writing and subsequent reading"},{"cell_type":"markdown","metadata":{},"source":"If you want to share this file with people who merely want to use it for visualization, a simple way to reduce the file size is by removing the dense scaled and corrected data matrix. The file still contains the raw data used in the visualizations in `adata.raw`."},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"adata.raw.to_adata().write(f\"{results_dir}/write/INCF_withoutX.h5ad\")"}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.5"}},"nbformat":4,"nbformat_minor":4}